diff --git master/app/controllers/application_controller.rb master/app/controllers/application_controller.rb
index 16a908c..60f4c56 100644
--- master/app/controllers/application_controller.rb
+++ master/app/controllers/application_controller.rb
@@ -24,9 +24,9 @@ class ApplicationController < ActionController::Base
   end
   protect_from_forgery
   
-  if SushiFabric::Application.config.fgcz?
-    before_action :authenticate_user!
-  end
+#  if SushiFabric::Application.config.fgcz?
+#    before_action :authenticate_user!
+#  end
   @@workflow_manager = DRbObject.new_with_uri(SushiFabric::WORKFLOW_MANAGER)
 
   def all_sushi_applications
diff --git master/app/helpers/application_helper.rb master/app/helpers/application_helper.rb
index 573029b..d58a1f5 100644
--- master/app/helpers/application_helper.rb
+++ master/app/helpers/application_helper.rb
@@ -9,7 +9,11 @@ module ApplicationHelper
       session[:projects] = if @fgcz and current_user
                              FGCZ.get_user_projects(current_user.login).map{|project| project.gsub(/p/,'').to_i}.sort
                            else
-                             [1001]
+#                             [1001]
+                              [
+                               1000, 1001, 1002, 1003, 1004, 1005, 1006, 1007, 1008, 1009, 1010,
+                               2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010
+                              ]
                            end
       session[:project] = if @fgcz and current_user
                             if project=params[:select_project] and number=project[:number] and number.to_i!=0 or
diff --git master/app/views/job_monitoring/index.html.erb master/app/views/job_monitoring/index.html.erb
index 297f98c..bd866af 100644
--- master/app/views/job_monitoring/index.html.erb
+++ master/app/views/job_monitoring/index.html.erb
@@ -20,9 +20,9 @@
 <% @job_list.each_with_index do |job, i| %>
   <tr>
     <td><%= job[0] %></td>
-    <td><%= button_to job[1], "/job_monitoring/#{job[0]}/change_status" %></td>
+    <td><%= job[1] %></td>
     <td>
-      <% if job[1]=='running' and current_user and job[4]==current_user.login %>
+      <% if job[1]=='running' %>
         <%= button_to 'kill', "/job_monitoring/#{job[0]}/kill_job", :confirm=>"Are you sure you want to kill the job '#{job[0]}'?" %>
       <% end %>
     </td>
diff --git master/app/views/shared/_header.html.erb master/app/views/shared/_header.html.erb
index 5714985..acc1dca 100644
--- master/app/views/shared/_header.html.erb
+++ master/app/views/shared/_header.html.erb
@@ -41,7 +41,6 @@
                     </ul>
                 </li>
             </ul>
-<% if user_signed_in? %>
 <% project_init %>
     <%= form_tag({:action => "index", :controller => "home"}, {:class => "navbar-form navbar-left", :style => "padding-top: 5px"}) do %>
         <div class="form-group">
@@ -55,11 +54,7 @@
     <% end %>
 
             <ul class="nav navbar-nav navbar-right list-inline" style="margin-right: 60px">
-		<li>
-		<%= link_to "Sign out", destroy_user_session_path, :method => :delete, :class => 'homelinks' %>
-		</li>
 	    </ul>
- <% end %>
 	</div>
         <!-- END menu navigation bar -->
     </div>
diff --git master/config/environments/production.rb master/config/environments/production.rb
index f7fbd1b..b491aea 100644
--- master/config/environments/production.rb
+++ master/config/environments/production.rb
@@ -79,12 +79,12 @@ SushiFabric::Application.configure do
   # fgcz
   if config.fgcz?
     #config.workflow_manager = "druby://fgcz-c-043:40001" # development
-    config.workflow_manager = "druby://fgcz-c-043:50001" # production
-    #config.workflow_manager = "druby://fgcz-c-043:70001" # demo
+    #config.workflow_manager = "druby://fgcz-c-043:50001" # production
+    config.workflow_manager = "druby://fgcz-c-043:70001" # demo
     config.scratch_dir = "/scratch"
     #config.gstore_dir = File.join(Dir.pwd, 'public/gstore/projects')
-    config.gstore_dir = "/srv/gstore/projects" # production
-    #config.gstore_dir = "/srv/GT/analysis/course_sushi/public/gstore/projects" # demo
+    #config.gstore_dir = "/srv/gstore/projects" # production
+    config.gstore_dir = "/srv/GT/analysis/course_sushi/public/gstore/projects" # demo
     #config.gstore_dir = "/srv/kenlab/shared_sib_sushi/gstore/projects" # sib_dev
     config.sushi_app_dir = Dir.pwd
     config.module_source = "/usr/local/ngseq/etc/lmod_profile"
diff --git master/lib/global_variables.rb master/lib/global_variables.rb
index 9373c7d..c4feaf6 100644
--- master/lib/global_variables.rb
+++ master/lib/global_variables.rb
@@ -2,8 +2,8 @@
 # encoding: utf-8
 
 GENOME_REF_DIR = '/srv/GT/reference'
-EZ_GLOBAL_VARIABLES = '/usr/local/ngseq/opt/EZ_GLOBAL_VARIABLES.txt'
-#EZ_GLOBAL_VARIABLES = '/usr/local/ngseq/opt/EZ_GLOBAL_VARIABLES_DEMO.txt'
+#EZ_GLOBAL_VARIABLES = '/usr/local/ngseq/opt/EZ_GLOBAL_VARIABLES.txt'
+EZ_GLOBAL_VARIABLES = '/usr/local/ngseq/opt/EZ_GLOBAL_VARIABLES_DEMO.txt'
 R_COMMAND = '/usr/local/ngseq/stow/R-3.3.2/bin/R'
 
 # compatible method to R vector, c(x,x,x) 
